File: daylist2.rb
require 'rspec'

class DayList
  include Enumerable

  def initialize(args)
    @start = args[:start]
    @end   = args[:end]

    @days_map = {
      sun:   0,
      mon:   1,
      tues:  2,
      wed:   3,
      thurs: 4,
      fri:   5,
      sat:   6,
    }

    @days = @days_map.keys
    @last_day = @days.size - 1
  end

  def each
    start_day = @days_map[@start]
    stop_day = @days_map[@end]

    is_overlap = (stop_day < start_day)

    last_day = if is_overlap
             @last_day
           else
             stop_day
           end

    (start_day..last_day).each do |day|
      yield @days[day]
    end

    if is_overlap
      (0..(start_day - 1)).each do |day|
        yield @days[day]
      end
    end
  end
end


describe DayList do
  it 'should exist' do
    @daylist = DayList.new(start: :sun, end: :sat)
    @daylist.should be_true
  end

  it 'should have a map function' do
    @daylist = DayList.new(start: :sun, end: :sat)
    list = @daylist.map {|d| d}
    list.length.should == 7
    list.should == [:sun, :mon, :tues, :wed, :thurs, :fri, :sat]
  end

  it 'should work starting in the middle of the week' do
    list = DayList.new(start: :wed, end: :tues).map {|d| d}
    list.length.should == 7
    list.should == [:wed, :thurs, :fri, :sat, :sun, :mon, :tues]
  end

  it 'should find beginning of list' do
    @daylist = DayList.new(start: :sun, end: :sat)
    start = @daylist.first
    start.should == :sun

  end

end

File: daylist2.pl
#!/usr/bin/perl

use strict;
use warnings;
use Data::Dumper;

my $days = {
    sun   => 0,
    mon   => 1,
    tues  => 2,
    wed   => 3,
    thurs => 4,
    fri   => 5,
    sat   => 6,
};

my $daylist = [qw/sun mon tues web thurs fri sat/];

print Dumper [ daylist( 'thurs', 'wed' ) ];

sub daylist {
    my ( $first, $last ) = @_;
    my $start = $days->{$first};
    my $end   = $days->{$last};
    if ( $start <= $end ) {
        return map { $daylist->[$_] } $start..$end;
    }
    else {
        my @list = map { $daylist->[$_] } $start..$#$daylist;
        push @list, daylist('sun', $last);
        return @list;
    }
}
